mmcv/ops/sync_bn.py

```diff
@@ -44,27 +44,48 @@ def forward(self, input, running_mean, running_var, weight, bias, momentum,
                    input, (torch.HalfTensor, torch.FloatTensor,
                            torch.cuda.HalfTensor, torch.cuda.FloatTensor)), \
                f'only support Half or Float Tensor, but {input.type()}'
-        output = torch.empty_like(input)
-        input3d = input.view(input.size(0), input.size(1), -1)
+        output = torch.zeros_like(input)
+        input3d = input.flatten(start_dim=2)
         output3d = output.view_as(input3d)
-
-        mean = torch.empty(
-            input3d.size(1), dtype=torch.float, device=input3d.device)
-        var = torch.empty(
-            input3d.size(1), dtype=torch.float, device=input3d.device)
-        norm = torch.empty_like(
+        num_channels = input3d.size(1)
+
+        # make sure mean/var/norm/std are initialized as zeros
+        # torch.empty does not guarantee that
+        mean = torch.zeros(
+            num_channels, dtype=torch.float, device=input3d.device)
+        var = torch.zeros(
+            num_channels, dtype=torch.float, device=input3d.device)
+        norm = torch.zeros_like(
             input3d, dtype=torch.float, device=input3d.device)
-        std = torch.empty(
-            input3d.size(1), dtype=torch.float, device=input3d.device)
+        std = torch.zeros(
+            num_channels, dtype=torch.float, device=input3d.device)
+
+        batch_size = input3d.size(0)
+        if batch_size > 0:
+            ext_module.sync_bn_forward_mean(input3d, mean)
+            batch_flag = torch.ones([1], device=mean.device, dtype=mean.dtype)
+        else:
+            # skip updating mean and leave it as zeros when the input is empty
+            # make sure there is gradient w.r.t input tensor
+            batch_flag = torch.zeros([1], device=mean.device, dtype=mean.dtype)
 
-        ext_module.sync_bn_forward_mean(input3d, mean)
+        # sync mean and the batch flag
+        vec = torch.cat([mean, batch_flag])
         if self.group_size > 1:
-            dist.all_reduce(mean, group=self.group)
-            mean /= self.group_size
-        ext_module.sync_bn_forward_var(input3d, mean, var)
+            dist.all_reduce(vec, group=self.group)
+            mean = vec[:num_channels] / self.group_size
+        # need total batch to decide whether to update the momentum
+        total_batch = vec[-1].detach().clamp(max=1)
```

Comment:
`total_batch_flag` may be better? In fact, this tensor works as a bool variable. Is it right?

Reply:
Actually yes, the momentum is either zero or the passed value for now.
Are you suggesting to make it a bool value or just rename it?

Reply:
Just renaming it is ok.

Reply:
OK